src/python/pants/core/goals/deploy.py

```diff
@@ -115,51 +128,294 @@ async def publish_process_for_target(
+
+        return None
+
+    @memoized_property
+    def output(self) -> tuple[str, ...]:
+        if self.outcome == _DeployStepOutcome.SKIPPED:
+            status = "skipped"
+            if self.description:
+                status += f" {self.description}"
+        else:
+            if self.outcome == _DeployStepOutcome.SUCCESS:
+                status = (
+                    "published" if self.action == _DeployStepAction.PUBLISH_PACKAGE else "deployed"
+                )
+                prep = "to"
+            else:
+                status = (
+                    "failed"
+                    if self.outcome == _DeployStepOutcome.FAILURE
+                    else "failed dependencies"
+                )
+                prep = "for"
+
+            if self.description:
+                status += f" {prep} {self.description}"
+
+        result = []
+        for name in self.names:
+            result.append(f"{name} {status}")
+
+        return tuple(result)
+
+    @property
+    def closure(self) -> Iterator[_DeployStepResult]:
+        for substep in self.substeps:
+            yield from substep.closure
+
+        yield self
+
+
+class _DeployStepResults(Collection[_DeployStepResult]):
+    @memoized_property
+    def exit_code(self) -> int:
+        if len(self) == 0:
+            return 0
+
+        return max(result.exit_code for result in self)
+
+    @property
+    def closure(self) -> Iterator[_DeployStepResult]:
+        for result in self:
+            yield from result.closure
+
+
+@rule(desc="Publish package dependencies", level=LogLevel.DEBUG)
+async def _run_publish_process(publish: PublishPackages) -> _DeployStepResult:
+    if publish.process:
+        # TODO cheating here, but only by invoking the underly process I can run this in a `@rule`
+        result = await Get(FallibleProcessResult, Process, publish.process.process)
+        return _DeployStepResult.from_publish_process_result(result, publish)
+
+    return _DeployStepResult.skipped(
+        action=_DeployStepAction.PUBLISH_PACKAGE,
+        names=publish.names,
+        description=publish.description,
+    )
+
+
+@rule(desc="Run deployment step", level=LogLevel.DEBUG)
+async def _run_deploy_process(deploy: DeployProcess) -> _DeployStepResult:
+    publish_results: Iterable[_DeployStepResult] = []
+    if deploy.publish_dependencies:
+        publish_processes = await _publish_processes_for_targets(deploy.publish_dependencies)
+        publish_results = _DeployStepResults(
+            await MultiGet(
+                Get(_DeployStepResult, PublishPackages, process) for process in publish_processes
+            )
+        )
+
+        if publish_results.exit_code != 0:
+            return _DeployStepResult.failed_dependencies(deploy, publish_results)
+
+    if deploy.process:
+        # TODO cheating here, but only by invoking the underly process I can run this in a `@rule`
+        result = await Get(FallibleProcessResult, Process, deploy.process.process)
```

Comment:
This is an important thing to nail down I think.

IIRC, the reason why we felt that `deployment` should be an interactive process was that some deploys require interactive password input or keychain access, etc. Not actually invoking these as interactive would break that. It would also potentially break caching I think, because `InteractiveProcess` isn't cached, while `Process` is, so this line will cache deployment processes.

BUT, in many cases a process does _not_ need to be interactive, and that allows things to run in parallel here, which is great.

I think that the best answer is likely something like #8923, as that would disconnect whether a process needs to run in the foreground from whether it can be run in parallel. I also don't think that that one would be that hard to do?

In the meantime though, could a python union type of `process: Process | InteractiveProcess` and then some conditional handling for "has any interactive" work...? Would likely be a pain.

Reply:
thanks for that, I also don't feel quite at ease to bypass the `InteractiveProcess` for something that is meant to be side-effecting, the caching itself is an important point as we would like to not cache in Pants the results of a deployment and leave it to the underlying tool to decide if a component needs a redeployment or not.

Will try using that union type and in case there are `InteractiveProcess` instances in the graph, run them sequentially before going for the parallel execution.